man/man1/pcp2graphite.1
.TH PCP2GRAPHITE 1 "PCP" "Performance Co-Pilot"
.SH NAME
.B pcp2graphite 
\- pcp-to-graphite data pump

.SH SYNOPSIS
.B pcp2graphite
[\f3\-h\f1 \f2HOST\f1]
[\f3\-a\f1 \f2ARCHIVE\f1]
[\f3\-L\f1]
[\f3\-O\f1 \f2TIMEORIGIN\f1]
[\f3\-T\f1 \f2TIMEFINISH\f1]
[\f3\-s\f1 \f2SAMPLES\f1]
[\f3\-t\f1 \f2DELTA\f1]
[\f3\-g\f1 \f2GRAPHITEHOST\f1]
[\f3\-p\f1 \f2PICKLEPORT\f1]
[\f3\-P\f1 \f2TEXTPORT\f1]
[\f3\-u\f1 \f2UNITS\f1]
[\f3\-m\f1 \f2PREFIX\f1]
\f2metricname\f1 ...

.SH DESCRIPTION
The program
.I pcp2graphite
is PCP monitor / data exporter program based on the Python PMAPI
bindings.  It relays whole hierarchies of the PMNS to a designated
Graphite (Carbon/Whisper) database by periodically polling, then
copying/converting relevant numerical metrics across the Graphite
"text" or "pickled" TCP protocols.  Metric values are optionally
scaled to a given unit/scale, but are not rate-converted.  Metric
names may be prefixed and are somewhat canonicalized to fit the
Graphite namespace rules.  One-shot conversion runs such as from
archives are not currently supported.

The data transfer will continue on a best-effort basis until the
program is interrupted.  Temporary errors are ignored.  Multiple
copies of this program may be run in parallel, to populate different
Graphite databases or subtrees, with different scaling or sampling
intervals.

Example of a systemd service file that can be used to have systemd 
managing a single instance of pcp2graphite:

.sp
.if n \{\
.RS 4
.\}
.nf
[Unit]
Description=Performance Co-Pilot data pump to Graphite
Documentation=man:pcp2graphite(8)
After=pmcd.service

[Service]
ExecStart=/opt/pcp/pcp2graphite.py <OPTIONS>

[Install]
WantedBy=multi\-user\&.target
.fi
.if n \{\
.RE
.\}
.PP

.SH OPTIONS

.TP
.B \-h HOST
Extract metrics from given PMCD server.
Default \f2local:\f1.

.TP
.B \-a ARCHIVE
Extract metrics from given archive.

.TP
.B \-L
Extract metrics from the local context PCP PMDAs.

.TP
.B \-t DELTA
Specify the polling interval in seconds.  If live, polling is aligned to
natural multiples of this interval.  Default \f260\f1

.TP
.B \-O TIMEORIGIN
Set the time origin for archive scanning.  The default is the start of the archive.

.TP
.B \-T TIMEFINISH
Set the time finish for archive scanning.  The default is the end of the archive.

.TP
.B \-s SAMPLES
Set the maximum number of samples to relay.  The default is unlimited.

.TP
.B \-g GRAPHITEHOST
Specify the graphite carbon-cache server where the relayed data is to
be relayed.  Default \f2localhost\f1.

.TP
.B \-p PICKLEPORT
Select the more efficient "pickled" encoding for carbon-cache
communication, and use the given TCP port.  This is default, using
port \f22004\f1.

.TP
.B \-P TEXTPORT
Alternately, select the less efficient "text" encoding for carbon-cache
communication, and use the given TCP port.  This protocol is not
default, because it requires a separate TCP connection for each
metric.  The typical port is \f22003\f1.

.TP
.B \-u UNITS
Rescale the metrics to the given UNITS string, as parsed by the
.I pmParseUnitsStr
function.  This must be dimensionally consistent with the underlying
PCP metric.  Default is no scaling.

.TP
.B \-m PREFIX
Use the given string as a prefix for all PMNS metrics relayed to Graphite.
Default is \f2pcp.\f1 (note the period at the end).

.TP
.B metricname ...
Relay all numeric metrics nested at or below the given name in the PMNS.
This option may be repeated.  This is a compulsory option, having no
default.

.PP
.SH "SEE ALSO"
.BR PCPIntro (1),
.BR python (1),
.BR pmcd (1),
.BR pmParseUnitsStr (3),
.BR http://graphite.wikidot.com/ ,
.BR pmval (1),
.BR PMAPI (3),
and
.BR pcp.conf (5).



CHANGELOG
pcp-3.10.8 (30 October 2015)
    - under development [see http://pcp.io/roadmap]

pcp-3.10.7 (16 September 2015)
    - containers: added the pcp-pmie container
    - containers: remove /proc bind mount, conflicts with --pid=host
    - pcp-atopsar: new PMAPI sar command based on http://atoptool.nl
    - pcp-uptime: fix sigsegv from missing metrics on first fetch
    - pmchart: fix memory corruption when restoring Saved Hosts list
    - pmchart: fix sigsegv opening a view when no archive available
    - rc_scripts: pmie/pmlogger - logged, no longer send cron email
    - rc_scripts: pmie/pmlogger - support for control.d directories;
      backward-compatible /etc/pcp/{pmie,pmlogger}/control extension
    - rc_scripts: pmcd/pmlogger - provide a localhost-only setting
    - pmdamic: new PMDA for Intel MIC card metrics
    - pmdaproc: export CPU scheduler cgroup metrics from cpu.stat
    - pmdalinux: export IPv6 kernel metrics from /proc/net/snmp6
    - pmdalinux: fix variable types for several disk metrics
    - pmdalinux: printf int64_t and/or uint64_t support (for softnet)
    - pmdajson: updates to allow use with python 2.6 installations
    - pmcpp: support inline shell command execution
    - pmwebd: multiple metrics/instances per log graphite mode speedup
    - libpcp: use public domain getdate instead of gnulib version
    - libpcp: optimise archive decompression for common operations
    - libpcp: revert pdubuf tsearch-based optimisation, fails on OSX
    - pmieconf: add dmthin pool utilization monitoring scripts
    - python: more care taken in pmOptions destructor (could sigsegv)
    - windows port: 64-bit format specifiers throughout code base
    - windows port: fix a libpcp dlsym bug and IPC on Windows
    - windows port: fix a python crash when free'ing resources
    - pmdawindows: fix mem.cache.bytes metric kernel extraction
    - perl pmda module: fix double free error in local_sock() routine
      (observed to cause pmdamemcache crash during error path exit)
    - pmdanginx: correct errors returned for not-running nginx daemon
    - pmdamounts: add mount fs usage metrics via statvfs(3)
    - pmdapostfix: improve error handling for files vs directories
    - pmdanfsclient: fix typos in metric value parsing routines
    - pmlogconf: remove tool-specific intervals from tool templates
    - pmlogconf: add nginx, elasticsearch and memcache configs
    - docs: update quickref guide for easier command pasting
    - docs: update references to init scripts, pmlogger control files
    - docs: update pmdanginx(1) describing nginx configuration needs
    - build: configure.ac preserves optional gcc flags in the build
    - build: cater for builds using the latest Vector (now 1.0.2)
    - build: fix LGPL license annotations in headers+libraries
    - build: current FreeBSD toolchain updates

pcp-3.10.6 (4 August 2015)
    - rc_scripts: use syslog over email for error notification
    - libpcp: catch bad valfmt encoding in interp.c (SGI BZ 1100)
    - libpcp: $PATH-alike lists for derived metrics configurations
    - libpcp: add instant() method for derived metrics
    - libpcp: fix locking problem in pmGetConfig(3) routine
    - python: fix bug in pmsubsys causing timestamp to reset.
    - misc: several Coverity scan bug fixes, across the board
    - containers: add version and arch to container image filenames
    - containers: add crontab entries for containers
    - containers: don't fix pcp user since pcp-base already does
    - pmlogcheck: teach it about valfmt badness
    - pmcpp: major revamp for pmlogger use (include files & macros)
    - pmcpp: speed and robustness enhancements
    - pmdadm: fix 2 dmcache metrics encoding on big-endian machines
    - pmdalinux: fix the mnt/net namespace setns call ordering
    - pmdalinux: corrections to the linux disk statistics metadata
    - pmdalinux: ipc shared memory metrics for total, rss, and swap
    - pmdalinux: new network.softnet metrics (/proc/net/softnet_stat)
    - pmdaproc: reduce log spam in pmdaproc unknown state case
    - pmdapostgresql: separate DB username from O/S username
    - pmdapostgresql: major overhaul for different versions schemas
    - pmdapostfix: general log parsing improvements
    - pmdajson: handle instances better in the JSON PMDA.
    - pmdajson: default to instantaneous metrics and do not attempt
      to guess at semantics from type.
    - pmdaperfevent: fix typo causing memleak/worse on an error path
    - pmdadtsrun: fix parenthesis botch in Perl code
    - pmdate: correct year and month offset problems at end of month
    - pmlogconf: stop enabling percpu interrupt metrics by default
    - pmlogger: improve handling of stale primary socket links
    - pmlogger: integration with pmcpp
    - pmwebd: elide repeated indom searching in graphite mode
    - pcp2graphite: archive mode, reused connections, and python3
    - pcp-atop: correct the reporting of some system memory metrics
    - pcp-python: allow preferred-python switching at runtime
    - pcp-collectl: fix an interrupt subsystem crash
    - pcp-collectl: break out append statements for valid syntax
    - pcp-collectl: fix recording of data to archive folios
    - iostat2pcp: fix off-by-one error in disk rawactive metrics
    - docs: corrections to pmdaperfevent man page
    - docs: add some missing text to pmGetConfig man page
    - docs: correct references to pre-v3 CC-BY-SA license
    - docs: update book builds for current Publican variants
    - docs: add derived metrics section to guide.html
    - rpms: pcp-collector meta-package default-enables pmcd+pmlogger
    - build: fix dependency problems causing parallel make races
    - build: pcp containers build support via Makepkgs
    - build: ensure pmcollectl is installed, man page updates
    - build: correct rpm package dependency on pcp-libs version
    - build: disable autofsd-probe on (most) platforms, unneeded
    - build: allow pmdajson configure on/off build overrides

src/pcp2graphite/pcp2graphite.py
#!/usr/bin/pcp python
#
# Copyright (C) 2014-2015 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
""" Relay PCP metrics to a graphite server """

import sys
import time


from pcp import pmapi
from cpmapi import PM_TYPE_32, PM_TYPE_U32, PM_TYPE_64, PM_TYPE_U64, PM_TYPE_DOUBLE
import cpmapi as c_api

class GraphiteRelay(object):
    """ Sends a periodic report to graphite about all instances of named metrics.
        Knows about some of the default PCP arguments.
    """

    def describeSource(self):
        """ Return a string describing our context; apprx. inverse of pmapi.fromOptions """

        ctxtype = self.context.type
        if ctxtype == c_api.PM_CONTEXT_ARCHIVE:
            return "archive " + ", ".join(self.opts.pmGetOptionArchives())
        elif ctxtype == c_api.PM_CONTEXT_HOST:
            hosts = self.opts.pmGetOptionHosts()
            if hosts is None: # pmapi.py idiosyncracy; it has already defaulted to this
                hosts = ["local:"]
            return "host " + ", ".join(hosts)
        elif ctxtype == c_api.PM_CONTEXT_LOCAL:
            hosts = ["local:"]
            return "host " + ", ".join(hosts)
        else:
            raise pmapi.pmUsageErr

    def __init__(self):
        """ Construct object, parse command line """
        self.context = None
        self.socket = None
        self.sampleCount = 0
        self.debug = False
        self.opts = pmapi.pmOptions()
        self.opts.pmSetShortOptions("a:O:s:T:g:p:P:u:m:t:h:t:D:LV?") # must include common options
        self.opts.pmSetShortUsage("[options] metricname ...")
        self.opts.pmSetOptionCallback(self.option)
        self.opts.pmSetOverrideCallback(self.option_override)
        # hack to include some explanatory text
        self.opts.pmSetLongOptionHeader("""
Description: Periodically, relay raw values of all instances of a
given hierarchies of PCP metrics to a graphite/carbon server on the
network.

Options""")
        self.opts.pmSetLongOptionVersion() # -V
        self.opts.pmSetLongOptionArchive() # -a FILE
        self.opts.pmSetLongOptionOrigin() # -O TIME
        self.opts.pmSetLongOptionSamples() # -s NUMBER
        self.opts.pmSetLongOptionFinish() # -T NUMBER
        self.opts.pmSetLongOptionDebug() # -D stuff
        self.opts.pmSetLongOptionHost() # -h HOST
        self.opts.pmSetLongOptionLocalPMDA() # -L
        self.opts.pmSetLongOptionInterval() # -t NUMBER
        self.opts.pmSetLongOption("graphite-host", 1, 'g', '', "graphite server host (default \"localhost\")")
        self.opts.pmSetLongOption("pickled-port", 1, 'p', '', "graphite pickled port (default 2004)")
        self.opts.pmSetLongOption("text-port", 1, 'P', '', "graphite plaintext port (usually 2003)")
        self.opts.pmSetLongOption("units", 1, 'u', '', "rescale all metric units (e.g., \"mbytes/5 sec\")")
        self.opts.pmSetLongOption("prefix", 1, 'm', '', "prefix for metric names (default \"pcp.\")")
        self.opts.pmSetLongOptionHelp()
        self.graphite_host = "localhost"
        self.graphite_port = 2004
        self.pickle = True
        self.prefix = "pcp."
        self.unitsstr = None
        self.units = None # pass verbatim by default
        self.units_mult = None # pass verbatim by default

        # now actually parse
        self.context = pmapi.pmContext.fromOptions(self.opts, sys.argv)
        self.interval = self.opts.pmGetOptionInterval() or pmapi.timeval(60,0)
        if self.unitsstr is not None:
            (self.units, self.units_mult) = self.context.pmParseUnitsStr(self.unitsstr)
        self.metrics = []
        self.pmids = []
        self.descs = []
        metrics = self.opts.pmNonOptionsFromList(sys.argv)
        if metrics:
            for m in metrics:
                try:
                    self.context.pmTraversePMNS(m, self.handle_candidate_metric)
                except pmapi.pmErr as error:
                    sys.stderr.write("Excluding metric %s (%s)\n" % (m, str(error)))
            sys.stderr.flush()

        if len(self.metrics) == 0:
            sys.stderr.write("No acceptable metrics specified.\n")
            raise pmapi.pmUsageErr()

        # Report what we're about to do
        print ("Relaying %d %smetric(s) with prefix %s from %s "
               "in %s mode to %s:%d every %f s" % (
                   len(self.metrics),
                   "rescaled " if self.units else "",
                   self.prefix,
                   self.describeSource(),
                   "pickled" if self.pickle else "text",
                   self.graphite_host, self.graphite_port,
                   self.interval))
        sys.stdout.flush()

    def option_override(self, opt):
        if (opt == 'p') or (opt == 'g'):
            return 1
        if (opt == 'D'): # pmapi.cc doesn't give us access to pmDebug, so intercept -D0
            self.debug = True
            # fallthrough
        return 0

    def option(self, opt, optarg, index): # need only handle the non-common options
        if (opt == 'g'):
            self.graphite_host = optarg
        elif (opt == 'p'):
            self.graphite_port = int(optarg if optarg else "2004")
            self.pickle = True
        elif (opt == 'P'):
            self.graphite_port = int(optarg if optarg else "2003")
            self.pickle = False
        elif (opt == 'u'):
            self.unitsstr = optarg
        elif (opt == 'm'):
            self.prefix = optarg
        else:
            raise pmapi.pmUsageErr()


    # Check the given metric name (a leaf in the PMNS) for
    # acceptability for graphite: it needs to be numeric, and
    # convertable to the given unit (if specified).
    #
    # Print an error message here if needed; can't throw an exception through the
    # pmapi pmTraversePMNS wrapper.
    def handle_candidate_metric(self, name):
        try:
            pmid = self.context.pmLookupName(name)[0]
            desc = self.context.pmLookupDescs(pmid)[0]
        except pmapi.pmErr as err:
            sys.stderr.write("Excluding metric %s (%s)\n" % (name, str(err)))
            return

        # reject non-numeric types (future pmExtractValue failure)
        types = desc.contents.type
        if not ((types == PM_TYPE_32) or
                (types == PM_TYPE_U32) or
                (types == PM_TYPE_64) or
                (types == PM_TYPE_U64) or
                (types == c_api.PM_TYPE_FLOAT) or
                (types == PM_TYPE_DOUBLE)):
            sys.stderr.write("Excluding metric %s (need numeric type)\n" % name)
            return

        # reject dimensionally incompatible (future pmConvScale failure)
        if self.units is not None:
            units = desc.contents.units
            if ((units.dimSpace != self.units.dimSpace) or
                (units.dimTime != self.units.dimTime) or
                (units.dimCount != self.units.dimCount)):
                sys.stderr.write("Excluding metric %s (incompatible dimensions)\n" % name)
                return

        self.metrics.append(name)
        self.pmids.append(pmid)
        self.descs.append(desc)


    # Convert a python list of pmids (numbers) to a ctypes LP_c_uint (a C array of uints).
    def convert_pmids_to_ctypes(self, pmids):
        import ctypes
        from ctypes import c_uint
        pmidA = (c_uint * len(pmids))()
        for i,p in enumerate(pmids):
            pmidA[i] = c_uint(p)
        return pmidA

    def send(self, timestamp, miv_tuples):
        import socket
        try:
            # reuse socket
            if self.socket is None:
                self.socket = socket.create_connection((self.graphite_host, self.graphite_port))
            if self.pickle:
                import pickle
                import struct
                pickled_input = []
                for (metric, value) in miv_tuples:
                    pickled_input.append((metric, (timestamp, value)))
                    # protocol=0 in case carbon is running under older python than we are
                    pickled_output = pickle.dumps(pickled_input, protocol=0)
                    header = struct.pack("!L", len(pickled_output))
                    msg = header + pickled_output
                    if (self.debug):
                        print ("Sending %s #tuples %d" %
                               (time.ctime(timestamp), len(pickled_input)))
                    self.socket.send(msg)
            else:
                for (metric, value) in miv_tuples:
                    message = ("%s %s %s\n" % (metric, value, timestamp))
                    msg = str.encode(message)
                    if (self.debug):
                        print ("Sending %s: %s" % (time.ctime(timestamp), msg.rstrip().decode()))
                    self.socket.send(msg)
        except socket.error as err:
            sys.stderr.write("cannot send message to %s:%d, %s, continuing\n" %
                             (self.graphite_host, self.graphite_port, err.strerror))
            self.socket = None
            return


    def sanitize_nameindom(self,str):
        """ Quote the given instance-domain string for proper digestion
        by carbon/graphite. """
        import re
        return "_" + re.sub('[^a-zA-Z_0-9-]','_', str)

    def execute(self):
        """ Using a PMAPI context (could be either host or archive),
            fetch and report a fixed set of values related to graphite.
        """

        # align poll interval to host clock
        if (self.context.type == c_api.PM_CONTEXT_HOST) or (self.context.type == c_api.PM_CONTEXT_LOCAL):
            time.sleep(float(self.interval) - (time.time() % float(self.interval)))

        # NB: we'd like to do: result = self.context.pmFetch(self.pmids)
        # But: pmFetch takes ctypes array-of-uint's, not a python list.  It needs to become polymorphic.
        result = self.context.pmFetch(self.convert_pmids_to_ctypes(self.pmids))
        sample_time = result.contents.timestamp.tv_sec # + (result.contents.timestamp.tv_usec/1000000.0)

        if (self.context.type == c_api.PM_CONTEXT_ARCHIVE):
            endtime = self.opts.pmGetOptionFinish()
            if endtime is not None:
                if (float(sample_time) > float(endtime.tv_sec)):
                    raise SystemExit

        miv_tuples = []

        for i, name in enumerate(self.metrics):
            for j in range(0,result.contents.get_numval(i)):
                # a fetch or other error will just omit that data value from the graphite-bound set
                try:
                    atom = self.context.pmExtractValue(
                        result.contents.get_valfmt(i),
                        result.contents.get_vlist(i,j),
                        self.descs[i].contents.type, c_api.PM_TYPE_FLOAT)

                    inst = result.contents.get_vlist(i,j).inst
                    if inst is None or inst < 0:
                        suffix=""
                    else:
                        suffix="."+self.sanitize_nameindom(self.context.pmNameInDom(self.descs[i], inst))

                    # Rescale if desired
                    if self.units is not None:
                        atom = self.context.pmConvScale(c_api.PM_TYPE_FLOAT, atom,
                                                        self.descs, i,
                                                        self.units)

                    if self.units_mult is not None:
                        atom.f = atom.f * self.units_mult

                    miv_tuples.append((self.prefix+name+suffix, atom.f))

                except pmapi.pmErr as error:
                    sys.stderr.write("%s[%d]: %s, continuing\n" % (name, inst, str(error)))
                    pass

        self.send(sample_time, miv_tuples)
        self.context.pmFreeResult(result)

        self.sampleCount += 1
        maxSampleCount = self.opts.pmGetOptionSamples()
        if maxSampleCount is not None and self.sampleCount >= maxSampleCount:
            raise SystemExit


if __name__ == '__main__':
    try:
        G=GraphiteRelay()
        while True:
            G.execute()

    except pmapi.pmUsageErr as usage:
        sys.stderr.write("\n")
        usage.message()
    except KeyboardInterrupt:
        pass
    except Exception as error:
        if str(error).find("PM_ERR_EOL") == -1:
            import traceback
            sys.stderr.write(str(error) + "\n") # init error: stop now
            sys.stderr.write(traceback.format_exc() + "\n")
